File: zombicideandroid/src/main/java/cc/game/zombicide/android/CLSendCommandSpinnerTask.java
package cc.game.zombicide.android;

import cc.lib.android.SpinnerTask;
import cc.lib.game.Utils;
import cc.lib.net.GameClient;
import cc.lib.net.GameCommand;
import cc.lib.net.GameCommandType;

class CLSendCommandSpinnerTask extends SpinnerTask<GameCommand> implements GameCommandType.Listener {

    final GameClient client;
    final ZClientMgr mgr;
    final GameCommandType responseType;

    public CLSendCommandSpinnerTask(ZombicideActivity context, GameCommandType responseType) {
        super(context);
        client = context.getClient();
        mgr = context.clientMgr;
        this.responseType = responseType;
        responseType.addListener(this);
    }

    @Override
    protected void doIt(GameCommand... args) throws Exception {
        client.sendCommand(args[0]);
        Utils.waitNoThrow(this, 20000);
    }

    @Override
    public void onCommand(GameCommand cmd) {
        synchronized (this) {
            notify();
        }
    }

    @Override
    protected void onCompleted() {
        responseType.removeListener(this);
    }
}

File: zombicideandroid/src/main/java/cc/game/zombicide/android/CLSendCommandSpinnerTask.kt
package cc.game.zombicide.android

import cc.lib.android.SpinnerTask
import cc.lib.net.GameCommand
import cc.lib.utils.KLock

open class CLSendCommandSpinnerTask(val context: ZombicideActivity) : SpinnerTask<GameCommand>(context), ZMPCommon.CLListener {
	private val clientMgr: ZClientMgr = context.clientMgr!!
	private val lock = KLock()

	override fun onPreExecute() {
		super.onPreExecute()
		clientMgr.addListener(this)
	}

	open val timeout = 10000L

	override suspend fun doIt(args: GameCommand?) {
		requireNotNull(context.clientMgr?.client).sendCommand(requireNotNull(args))
		if (timeout > 0)
			lock.acquireAndBlock(timeout) {
				throw Exception("timeout")
			}
		else
			lock.acquireAndBlock()
	}

	fun release() {
		lock.release()
	}

	override fun onCancelButtonClicked() {
		lock.release()
	}

	override fun onCompleted() {
		clientMgr.removeListener(this)
	}

	override fun onNetError(e: Exception) {
		onError(e)
	}

}